pkg/kubecost/allocationfilter.go

```diff
@@ -0,0 +1,210 @@
+package kubecost
+
+import "github.com/kubecost/cost-model/pkg/log"
+
+// FilterField is an enum that represents Allocation-specific fields that can be
+// filtered on (namespace, label, etc.)
+type FilterField string
+
+// If you add a FilterField, MAKE SURE TO UPDATE ALL FILTER IMPLEMENTATIONS! Go
+// does not enforce exhaustive pattern matching on "enum" types.
+const (
+	FilterClusterID      FilterField = "clusterid"
+	FilterNode                       = "node"
+	FilterNamespace                  = "namespace"
+	FilterControllerKind             = "controllerkind"
+	FilterControllerName             = "controllername"
+	FilterPod                        = "pod"
+	FilterContainer                  = "container"
+
+	// Filtering based on label aliases (team, department, etc.) should be a
+	// responsibility of the query handler. By the time it reaches this
+	// structured representation, we shouldn't have to be aware of what is
+	// aliased to what.
+
+	FilterLabel      = "label"
+	FilterAnnotation = "annotation"
+)
+
+// FilterOp is an enum that represents operations that can be performed
+// when filtering (equality, inequality, etc.)
+type FilterOp string
+
+// If you add a FilterOp, MAKE SURE TO UPDATE ALL FILTER IMPLEMENTATIONS! Go
+// does not enforce exhaustive pattern matching on "enum" types.
+const (
+	FilterEquals    FilterOp = "equals"
+	FilterNotEquals          = "notequals"
+)
+
+// AllocationFilter represents anything that can be used to filter an
+// Allocation.
+//
+// Implement this interface with caution. While it is generic, it
+// is intended to be introspectable so query handlers can perform various
+// optimizations. These optimizations include:
+// - Routing a query to the most optimal cache
+// - Querying backing data stores efficiently (e.g. translation to SQL)
+//
+// Custom implementations of this interface outside of this package should not
+// expect to receive these benefits. Passing a custom implementation to a
+// handler may in errors.
+type AllocationFilter interface {
+	// Matches is the canonical in-Go function for determing if an Allocation
+	// matches a filter.
+	Matches(a *Allocation) bool
+}
```

Comment:
I like the flexibility here. We'll just want to be careful with implementation -- i.e. if the `Query` functions take this as a type, they still won't be introspectable (is that a word?) without type casting.

Reply:
Do you have a sense for what type, precisely, we'd want to use for `QueryOptions`?

Reply:
The need to type cast is unfortunate. I think its inevitable that `QueryOptions` will have a `Filter kubecost.AllocationFilter` field, though -- filters can be nested within ANDs and ORs and I can't think of a good way to represent that general idea without an interface.

The fallback plan is fairly straightforward, I think. If whatever takes `QueryOptions` isn't able to introspect (due to an unknown implementation of `AllocationFilter` OR a bug in introspection) it can just fall through to calling `Matches()` at the end of the day.

I miss ML-style match statements :(

Reply:
> filters can be nested within ANDs and ORs and I can't think of a good way to represent that general idea without an interface

This was actually part of my reasoning about making it a single "composite" filter with an `Op` -- so that you could express multiple kinds of composite filters without an interface. But trust your design decisions! I can see it either way.